### PR TITLE
add metric for mnist.

### PR DESCRIPTION
**Pull request by heavengate**
**add metric for mnist**
- Add `Accuracy` metric for mnist and adapt for static and dygraph
- **train/eval mode only return `loss` and `metric`, test mode only return `outputs`**
- `YOLOv3` metric input and loss input is different, consider to add as a callback later?

### CODE REVIEW

File: metrics/coco.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+logging.basicConfig(level=logging.INFO, format=FORMAT)
+logger = logging.getLogger(__name__)
+
+__all__ = ['COCOMetric']
+
+
+OUTFILE = './bbox.json'
+
+
+class COCOMetric(Metric):
```
Comment (qingqing01) [inline metrics/coco.py]: 1. Better to add  `Accuracy` as the first metric.
2. The built-in metrics in framework should not contain COCO, since it only works for detection task and depends on cocoapi.

Comment (heavengate) [reply metrics/coco.py]: Done, thanks!

File: model.py [hunk 2]
```diff
@@ -277,15 +298,26 @@ def _run(self, inputs, labels=None, device='CPU', device_ids=None):
...
+        # LoDTensor cannot be fetch as numpy directly
+        rets = [np.array(v) for v in rets]
+        if self.mode == 'test':
+            return rets[:]
+        losses = rets[:num_loss]
+        metric_states = restore_flatten_list(rets[num_loss:], metric_splits)
+        metrics = []
+        for metric, state in zip(self.model._metrics, metric_states):
+            metrics.append(metric.update(*state))
+        return losses, metrics
```
Comment (guoshengCS) [inline model.py]: Maybe we should return metrics optionally  according to whether use metrics or not.

File: model.py [hunk 3]
```diff
@@ -422,10 +462,14 @@ def eval(self, inputs, labels, device='CPU', device_ids=None):
...
+        outputs = to_list(self.model.forward(*[to_variable(x) for x in inputs]))
         losses = self.model._loss_function(outputs, labels)
-        return [to_numpy(o) for o in to_list(outputs)], \
-            [to_numpy(l) for l in losses]
+        metrics = []
+        for metric in self.model._metrics:
+            metric_outs = metric.add_metric_op(outputs, [to_variable(l) for l in labels])
+            m = metric.update(*[to_numpy(m) for m in to_list(metric_outs)])
+            metrics.append(m)
+        return [to_numpy(l) for l in losses], metrics
```
Comment (guoshengCS) [inline model.py]: Same as above.

File: model.py (hunk 2 above)
Comment (heavengate) [reply model.py]: Done, thanks!

File: model.py (hunk 3 above)
Comment (heavengate) [reply model.py]: Done, thanks!

File: model.py [hunk 4]
```diff
@@ -429,32 +464,43 @@ def train(self, inputs, labels=None):
...
         losses = self.model._loss_function(outputs, labels)
         final_loss = fluid.layers.sum(losses)
         final_loss.backward()
         self.model._optimizer.minimize(final_loss)
         self.model.clear_gradients()
-        return [to_numpy(o) for o in to_list(outputs)], \
-            [to_numpy(l) for l in losses]
+        metrics = []
+        for metric in self.model._metrics:
+            metric_outs = metric.add_metric_op(outputs, [to_variable(l) for l in labels])
```
Comment (LielinJiang) [inline model.py]: label的to_variable可以再调用loss之前。目前的写法应该是重复拷贝了。

Comment (heavengate) [reply model.py]: Done, thanks
